app/javascript/components/util/DateInput.js

```diff
@@ -3,34 +3,85 @@ import PropTypes from 'prop-types';
 import moment from 'moment';
+import _ from 'lodash';
 
 class DateInput extends React.Component {
   constructor(props) {
     super(props);
     this.datePickerRef = React.createRef();
-    this.handleRawChange = this.handleRawChange.bind(this);
-    this.handleDateChange = this.handleDateChange.bind(this);
-    this.clearDate = this.clearDate.bind(this);
+    this.state = {
+      currentDate: props.date || null,
+      lastValidDate: props.date || null,
+    };
   }
 
-  handleDateChange(date) {
-    this.props.onChange(date && moment(date).format('YYYY-MM-DD'));
+  /**
+   * Called only when date value is changed
+   */
+  handleDateChange = date => {
+    const momentDate = date && moment(date).format('YYYY-MM-DD');
+    this.props.onChange(momentDate);
     this.datePickerRef.current.setOpen(false);
-  }
+    if (this.dateIsValidAndNotEmpty(momentDate)) {
+      this.setState({ lastValidDate: momentDate });
+    }
+  };
 
-  handleRawChange(event) {
-    if (event.target.value) {
-      this.datePickerRef.current.setOpen(true);
+  /**
+   * Called when typing into the date input
+   */
+  handleRawChange = event => {
+    this.datePickerRef.current.setOpen(true);
+    this.setState({ currentDate: event.target.value });
+  };
+
+  /**
+   * Called when day is clicked in the datepicker
+   */
+  handleSelect = date => {
+    this.setState({ currentDate: date });
+  };
+
+  /**
+   * Called when the date input goes out of focus.
+   * This will happen when the user leaves the form input (i.e. clicks out of the datepicker)
+   */
+  handleOnBlur = () => {
+    const currentDate = this.state.currentDate && moment(this.state.currentDate).format('YYYY-MM-DD');
+    // if date is not valid when clicking out of the date input
+    if (!this.dateIsValidAndNotEmpty(currentDate)) {
+      // change back to the last valid date or today if the field should revert to the last valid date
+      if (this.props.replaceBlank) {
+        const date = this.state.lastValidDate || moment().format('YYYY-MM-DD');
+        this.props.onChange(date);
+        // clear the date if field should be empty on invalid
+      } else if (this.props.clearInvalid) {
+        this.clearDate();
+      }
     }
-  }
+  };
 
-  clearDate() {
-    event.preventDefault();
-    this.props.onChange(null);
-    this.datePickerRef.current.setOpen(false);
-  }
+  /**
+   * Returns false if date is null, undefined, invalid or falls outside the min/max date range
+   * Returns true otherwise
+   */
+  dateIsValidAndNotEmpty = date => {
+    const isNotNull = !_.isNil(date);
+    const isValid = moment(date, 'YYYY-MM-DD').isValid();
+    const isInRange = moment(this.props.minDate).isBefore(date) && moment(this.props.maxDate).isAfter(date);
+    return isNotNull && isValid && isInRange;
+  };
+
+  /**
+   * Clears the selected date in parent component and closes the datepicker
+   */
+  clearDate = () => {
+    this.setState({ currentDate: null }, () => {
+      this.props.onChange(null);
```

Comment:
is there a reason why this was moved to the callback of `setState`? these aren't dependent on `state`, so I think this is necessary

Reply:
got catch, updated here: https://github.com/SaraAlert/SaraAlert/pull/734/commits/3f3366ba02e8cc33155da0d952ce26786b27c687